compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3432,32 +3432,32 @@ public static void fixSelfReferencingSameUnion(BType originalMemberType, BUnionT
 
     private boolean isFiniteTypeAssignable(BFiniteType finiteType, BType targetType, Set<TypePair> unresolvedTypes) {
         BType expType = getReferredType(targetType);
-        if (expType.tag == TypeTags.FINITE) {
-            return finiteType.getValueSpace().stream()
-                    .allMatch(expression -> isAssignableToFiniteType(expType, (BLangLiteral) expression));
-        }
+        for (BLangExpression expression : finiteType.getValueSpace()) {
+            if (expType.tag != TypeTags.FINITE && expType.tag != TypeTags.UNION) {
+                if (!isLiteralCompatibleWithBuiltinTypeWithSubTypes((BLangLiteral) expression, targetType) &&
+                        !isAssignable(expression.getBType(), expType, unresolvedTypes)) {
+                    return false;
+                }
+            }
 
-        if (targetType.tag == TypeTags.UNION) {
-            List<BType> unionMemberTypes = getAllTypes(targetType, true);
-            for (BLangExpression valueExpr : finiteType.getValueSpace()) {
+            ((BLangLiteral) expression).isFiniteContext = true;
+            if (expType.tag == TypeTags.FINITE) {
+                boolean foundMember = isAssignableToFiniteType(expType, (BLangLiteral) expression);
+                if (!foundMember) {
```

Comment:
We do not need temp variable here

Reply:
Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/35151/commits/76e1ffa5d1172659625f9c177cb782ab4c47d307